### PR TITLE
feat(poseidon-cipher-circuit): implement Poseidon Cipher circuits in circom

### LINKED ISSUE #121: Implement Poseidon Decryption circuits in circom
**Additional context**

Currently we have the typescript implementation, but no circom counterpart.

### PR DESCRIPTION
**Pull request by ctrlc03**
## Description

Implement decryption circuits for Poseidon-cipher and associated tests.

## Related Issue

fix #121 

## Does this introduce a breaking change?

-   [ ] Yes
-   [x] No

### CODE REVIEW

File: packages/circuits/circom/poseidon-cipher.circom [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    // Using recommended parameters from whitepaper https://eprint.iacr.org/2019/458.pdf (table 2, table 8)
+    // Generated by https://extgit.iaik.tugraz.at/krypto/hadeshash/-/blob/master/code/calc_round_numbers.py
+    // And rounded up to nearest integer that divides by t
+    var N_ROUNDS_P[16] = [56, 57, 56, 60, 60, 63, 64, 63, 60, 66, 60, 65, 70, 60, 64, 68];
+    var nRoundsF = 8;
+    var nRoundsP = N_ROUNDS_P[nInputs - 2];
+    var C[nInputs*(nRoundsF + nRoundsP)] = POSEIDON_C_OLD(nInputs);
+    var M[nInputs][nInputs] = POSEIDON_M_OLD(nInputs);
+
+    component ark[nRoundsF + nRoundsP];
```
Comment (0xjei) [inline packages/circuits/circom/poseidon-cipher.circom:178]: Just a note that this part of the code can be improved for readability in the future.

File: packages/circuits/package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@zk-kit/circuits",
-    "version": "0.2.4",
+    "version": "0.2.5",
```
Comment (0xjei) [inline packages/circuits/package.json:3]: @cedoor are you okay with this bump?

File: packages/circuits/tests/common.ts [hunk 3]
```diff
@@ -10,3 +13,55 @@ export const circomkit = new Circomkit({
...
+/**
+ * Returns a BabyJub-compatible random value. We create it by first generating
+ * a random value (initially 256 bits large) modulo the snark field size as
+ * described in EIP197. This results in a key size of roughly 253 bits and no
+ * more than 254 bits. To prevent modulo bias, we then use this efficient
+ * algorithm:
+ * http://cvsweb.openbsd.org/cgi-bin/cvsweb/~checkout~/src/lib/libc/crypt/arc4random_uniform.c
+ * @returns A BabyJub-compatible random value.
+ */
+export const genRandomBabyJubValue = (): bigint => {
```
Comment (0xjei) [inline packages/circuits/tests/common.ts:26]: Another note: these methods are duplicated from @zk-kit/poseidon-cipher test utils. We should move things in a common space in the future.

**Review by 0xjei — APPROVED** [review]
Great job @ctrlc03 :fire: The implementation is consistent with the TS one in `poseidon-cipher` which follows the paper :+1: The separation of concerns at templates level enhances circuit readability. The focus of this PR is not on improvement, so we can work on them in the future.

File: packages/circuits/circom/poseidon-cipher.circom (hunk 1 above)
Comment (ctrlc03) [reply packages/circuits/circom/poseidon-cipher.circom:178]: 101%!!

File: packages/circuits/package.json (hunk 2 above)
Comment (cedoor) [reply packages/circuits/package.json:3]: Umh I think it should be at least a minor in this case, as there are additional circuits, not patches.

File: packages/circuits/tests/common.ts (hunk 3 above)
Comment (cedoor) [reply packages/circuits/tests/common.ts:26]: `@zk-kit/utils` sounds like a good place to move them IMO.

We can add a task for this.

File: packages/circuits/package.json (hunk 2 above)
Comment (ctrlc03) [reply packages/circuits/package.json:3]: 0.3.0 then?

Comment (cedoor) [reply packages/circuits/package.json:3]: Yes, sounds great!

Comment (ctrlc03) [reply packages/circuits/package.json]: Done

**Review by cedoor — APPROVED** [review]
I haven't gone super deep but it looks great to me! Great work @ctrlc03 and thanks a lot for this contribution 🙏🏽

**Pull request merged**